azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -1256,76 +1361,287 @@ private void fillUpdateTimeAndExecId(List<ExecutableFlow> flows,
-      return host;
+    public void setActive(boolean isActive) {
+      this.isActive = isActive;
     }
 
-    public int getPort() {
-      return port;
+    public boolean isActive() {
+      return isActive;
     }
 
-    @Override
-    public int hashCode() {
-      final int prime = 31;
-      int result = 1;
-      result = prime * result + ((host == null) ? 0 : host.hashCode());
-      result = prime * result + port;
-      return result;
+    @SuppressWarnings("unused")
+    public void shutdown() {
+      shutdown = true;
+      this.interrupt();
     }
 
-    @Override
-    public boolean equals(Object obj) {
-      if (this == obj)
-        return true;
-      if (obj == null)
-        return false;
-      if (getClass() != obj.getClass())
-        return false;
-      ConnectionInfo other = (ConnectionInfo) obj;
-      if (host == null) {
-        if (other.host != null)
-          return false;
-      } else if (!host.equals(other.host))
-        return false;
-      if (port != other.port)
-        return false;
-      return true;
+    public void run() {
+      while (!shutdown) {
+        synchronized (this) {
+          try {
+            long currentTime = System.currentTimeMillis();
+            if (isActive
+              && currentTime - QUEUE_PROCESSOR_WAIT_IN_MS > lastProcessingTime) {
+              // Refresh executor stats to be used by selector
+              refreshExecutors();
+              // process upto a maximum of maxContinousSubmission from queued
+              // flows
+              processQueuedFlows(maxContinousSubmission);
+              lastProcessingTime = currentTime;
+            }
+            wait(QUEUE_PROCESSOR_WAIT_IN_MS);
+          } catch (InterruptedException e) {
+            logger.info("Interrupted. Probably to shut down.");
+          }
+        }
+      }
     }
   }
 
-  @Override
-  public int getExecutableFlows(int projectId, String flowId, int from,
-      int length, List<ExecutableFlow> outputList)
-      throws ExecutorManagerException {
-    List<ExecutableFlow> flows =
-        executorLoader.fetchFlowHistory(projectId, flowId, from, length);
-    outputList.addAll(flows);
-    return executorLoader.fetchNumExecutableFlows(projectId, flowId);
+  private void refreshExecutors() {
+    synchronized (activeExecutors) {
+      // TODO: rest api call to refresh executor stats
+    }
   }
 
-  @Override
-  public List<ExecutableFlow> getExecutableFlows(int projectId, String flowId,
-      int from, int length, Status status) throws ExecutorManagerException {
-    return executorLoader.fetchFlowHistory(projectId, flowId, from, length,
-        status);
+  private void processQueuedFlows(long maxContinousSubmission) {
+    try {
+      Pair<ExecutionReference, ExecutableFlow> runningCandidate;
+      int submissionNum = 0;
+      while (submissionNum < maxContinousSubmission
+        && (runningCandidate = queuedFlows.peek()) != null) {
+        synchronized (runningCandidate) {
```

Comment:
I think we will still have racing condition if only synchronize on the runningCandidate.  

think the scenario - 
two threads hit this simultaneously, both passed the peek and got the same object reference,  then one thread passes into the synchronized block and one waiting,  thread #1 processes the flow, assign executor , put it in the running queue then exit the block, then thread #2 kicks, do all the work again (which may assign a different executor) and put again the flow in the running flow list, which is dup , which is not what we want.  

Let me know if I got it right ,  I think we should be locking on the queuedFlows object instead of  the runningcandidate.